epsagon/utils.py

```diff
@@ -445,3 +456,123 @@ def database_connection_type(hostname, default_type):
+    :var bounds: token_type is sqlparse type
+    :var bounds: leftmost bound and rightmost bound are absolute limits
+    :var bounds: separator is the symbol(s) between values
+    :var bounds: signal is a preceding identifier for each value
+    """
+    bounds = {
+        'select': [
+            ('token_type', sqlparse.sql.Where),
+            ('left_bound', 'WHERE '),
+            ('right_bound', None),
+            ('separators', [')', ' AND ', ' OR ']),
+            ('operators', ['>=', '<=', '>', '<', '=', '<>'])
+        ],
+        'insert': [
+            ('token_type', sqlparse.sql.Values),
+            ('left_bound', 'VALUES ('),
+            ('right_bound', ')'),
+            ('separators', ', '),
+            ('operators', None),
+        ],
+
+    }
+
+    if isinstance(query, bytes):
+        query = query.decode('UTF-8')
+
+    if isinstance(operation, bytes):
+        operation = operation.decode('UTF-8')
+
+    if operation not in bounds.keys():
+        return query
+
+    # order not preserved in py2. Dont default to OrderedDict,not inefficient
+    to_dict = collections.OrderedDict if sys.version_info.major == 2 else dict
+
+    [token_type,
+    start, stop,
+    separators, operators] = to_dict(bounds[operation]).values()
+    obfuscated_query = []
+
+    parse_query = sqlparse.parse(query)[0].tokens
+
+    for token in parse_query:
+        t = str(token)
+
+        if isinstance(token, token_type):
+
+            # indices of start and stop positions
+            positions = (
+                t.upper().find(start) + len(start) if start else -1,
+                t.upper().find(stop) if stop else len(t) - 1,
+            )
+
+            if -1 in positions or positions[0] > positions[1]:
+                print_debug('could not obfuscate clause: {clause}'
+                            .format(clause=token_type))
+                continue
+
+            values = t[positions[0]:positions[1]]
+
+            separators = build_split_string(separators)
+            operators = build_split_string(operators)
+
+            values = re.split(separators, values, flags=re.IGNORECASE) or []
+
+            # step ignores the seperators in between
+            for i, v in enumerate(values[::2]):
+                i *= 2  # enumerate properly for step
+
+                # partition into 3 if operator defined, mask the right
+                if operators:
+                    v = re.split(operators, v, flags=re.IGNORECASE)
+                    if len(v) != 3:
+                        continue
+
+                    v[0] = v[0].strip()
+                    v[2] = OBFUSCATION_MASK
+                    v = ''.join(v)
+                # else replace entire value
+                else:
+                    v = OBFUSCATION_MASK
+                values[i] = v
+
+            values = ''.join(values)
+
+            t = t[:positions[0]] \
+                + values \
+                + t[positions[1]:]
+
+        obfuscated_query.append(t)
+
```

Comment:
try to split this function as it became too long :) (the internal logic for example can be moved)